presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java

```diff
@@ -1457,6 +1456,69 @@ protected Node aggregateResult(Node aggregate, Node nextResult)
         throw new UnsupportedOperationException("not yet implemented");
     }
 
+    private static String decodeStringExpression(SqlBaseParser.StringContext context)
+    {
+        String rawContent = unquote(context.STRING().getText());
+        if (context.ESCAPE_INTRODUCER() == null) {
+            return rawContent;
+        }
+        String result = "";
+        String currentEscapedCode = "";
+        int lengthBeforeComplete = 0;
+        for (int i = 0; i < rawContent.length(); i++) {
+            if (rawContent.charAt(i) == '\\') {
+                if (lengthBeforeComplete != 0) {
+                    throw parseError("Incomplete escaped character: " + currentEscapedCode, context);
+                }
+                if (i < rawContent.length() - 1 && rawContent.charAt(i + 1) == '+') {
+                    lengthBeforeComplete = 6;
+                    i++;
+                }
+                else {
+                    lengthBeforeComplete = 4;
+                }
+                continue;
+            }
+            if (lengthBeforeComplete != 0) {
+                if (!isHexDigit(rawContent.charAt(i))) {
+                    throw parseError("Invalid hexadecimal digit: " + rawContent.charAt(i), context);
+                }
+                currentEscapedCode += rawContent.charAt(i);
+                lengthBeforeComplete--;
+            }
+            if (lengthBeforeComplete == 0) {
+                if (currentEscapedCode.isEmpty()) {
+                    // A non-escaped character
+                    result += rawContent.charAt(i);
+                }
+                else {
+                    int codePoint = Integer.parseInt(currentEscapedCode, 16);
+                    if (!Character.isValidCodePoint(codePoint)) {
+                        throw parseError("Invalid escaped character: " + currentEscapedCode, context);
+                    }
+                    if (Character.isSupplementaryCodePoint(codePoint)) {
+                        char[] temp = Character.toChars(codePoint);
+                        result += temp[0];
+                        result += temp[1];
+                    }
+                    else {
+                        char currentCodePoint = (char) codePoint;
+                        if (Character.isSurrogate(currentCodePoint)) {
```

Comment:
I assume a single surrogate cannot be greater than 16 bit right? So if we cast a 6-digit escaped character into an integer, it will always be a non-supplementary code point. BTW Character::isSurrogate can only take 'char' as an input. Added a test case '\+00DBFF' to cover this.